### ISSUE TITLE
Cancel git Sync's pull (and the sync acttion itself) if behind the remote

### ISSUE THREAD

**Issue by Gouvernathor** [microsoft/vscode#174198]
Twice already, I have tried to push commits to a repo using the Sync button, only to find out after the fact that I was behind the remote, and that instead of cancelling the operation when the fetch ended up like that, VSCode decided by itself to pull and merge my commits with the distant one and push the merged timeline. Once, it made the repo into a bogus state and we needed to force-push to fix it.

When I was on Atom, there was no such thing. There was a main git button, just like Sync, but when you were trying to push and the fetch returned that you were behind the remote, the whole thing was cancelled and returned to the user to decide what to do.
Please be like Atom.

**Comment by sergio-asenjo** [comment]
Any news on this? I've been having the exact same problem. I'm used to use the sync button in the git section, but due to the extra merge, it's becoming annoying.

**Comment by lszomoru** [comment]
@sergio-asenjo, could you please describe your scenario in more details? Thank you!

**Comment by albertosantini** [comment]
Maybe I am wrong, but that is git default behaviour when you pull code.

Generally speaking in your `.gitconfig` you should change the default:
`git config --global pull.rebase true`

**Comment by Gouvernathor** [comment]
@lszomoru which details do you think are missing from my original scenario ?

@albertosantini we are talking about the behavior of push, not pull.

**Comment by albertosantini** [comment]
@Gouvernathor `VSCode sync = git pull + git push`

If the local is behind the remote, the default strategy is merging; therefore the merged history is pushed.
You can see the command behind the scenes in git ouput of VSCode.

**Comment by Gouvernathor** [comment]
Exactly, that's the problem. It should never push a commit (the merge commit) which the user never had the occasion to review.

**Comment by albertosantini** [comment]
Generally speaking, the actual behaviour fits in our workflow (with rebase strategy configured).

I was wondering, before clicking, if you give a look at those numbers close to "Syncronize changes": they display behind and local commits; so you are aware if you are behind.

At last, to mitigate your scenario, I suggest these VSCode settings:

    "git.autofetch": true,
    "git.rebaseWhenSync": true,
    "git.confirmSync": true,

Hope that helps.

**Comment by nicomunting** [comment]
Changing this setting is indeed helpful, @albertosantini. I still think that the default behavior of VSCode in this case is not friendly and agree with the OP that the user should have a chance to review before a merge commit is being pushed on the branch.

**Comment by Gouvernathor** [comment]
@albertosantini I'm not sure what the "rebase" setting has to do with the situation : if and when there is a conflict, the question is whether or not to resolve it without user input (and then whether or not to push the result to the remote without user input). The question of whether the conflict is resolved by a classic merge, by a rebase, or even by a squash-merge, is not really the matter.
Unless the "rebaseWhenSync" option isn't really about rebase vs. merge ?

The autofetch I already knew, it's a good thing (and probably should be the default) but unless it runs every couple of seconds, it can't completely prevent a situation where a fetch finds remote commits to exist _after_ the user clicks the sync button, it only lessens that probability.

Finally, the "confirm sync" : if it's a systematic confirmation pop-up that doesn't depend on the bullet points I listed earlier, then it's just something bothersome that users will learn to confirm without reading. It will only slow people down and not prevent any of the problems that occurred.

**Comment by Gouvernathor** [comment]
The behavior I seek is roughly as follows :
```py
local_unpushed_commits: bool
remote_incoming_commits: bool
def git_sync():
    if local_unpushed_commits:
        if remote_incoming_commits:
            git.pull()
        else:
            remote_incoming_commits = git.fetch()
            if remote_incoming_commits:
                pass # possibly pull, but never push
            else:
                git.push()
    else:    
        git.pull()
```

**Comment by lszomoru** [comment]
Thanks for the feedback! I did not see it mentioned here, so please note that you can hide the "Sync Changes" action button using the `git.showActionButton` setting. After hiding the button you can manually invoke the "Git: Pull" and "Git: Push" commands to have better control.

Rather than tweaking the logic behind the "Sync Changes", I think that we should provide users with the option to have "Pull" and "Push" buttons. If we know that there are incoming changes, we would only show "Pull" and users would have to manually pull the changes before they can push their changes to the remote.

**Comment by Gouvernathor** [comment]
@lszomoru using pull and push separately is what I'm doing ever since. I'll take the option to hide Sync, thanks.

I think I agree that keeping a button named "Sync" and have it change behavior would disrupt people's use of VSCode.
I don't think it makes much sense to have both a Push and a Pull buttons show at the same time : Pull only makes sense when we know there are remote commits, Push only when there are local unpushed commits. When there are both local unpushed and remote commits, well, the Push button cannot act - unless you want it to force-push by default, but I think we can agree that it's a bad thing, and it should be done as purposefully as possible.
So, I think it only makes sense to have one of the Push and Pull buttons show at a given time. And when there are neither local unpushed commits nor remote incoming commits, well, why not have a Fetch button ?
...That brings me to the behavior I described above, if you read Python.

So, how about having a button that _replaces_ the Sync button (based upon configuration variables and/or compatibility stuff), and which shows Pull whenever there are remote incoming commits, else Push if there are unpushed local commits, else Fetch ?
And have the Push abort itself if there happens to be remote commits that we weren't aware of when the button was clicked.
I honestly don't see what drawback you could see in such a feature, if you set aside compatibility and transition concerns.

**Comment by alexrecuenco** [comment]
If you want to make sure that pulling doesn't do a merge commit

```
pull.ff=only
```

```bash
git config pull.ff only
```

Once you do that if you use the "Sync" feature and it has issues, a pop up with an error will appear saying

```txt
> git pull --tags origin renovate/configure
From ssh://github.com/myuser/myrepo
* branch            renovate/configure -> FETCH_HEAD
hint: Diverging branches can't be fast-forwarded, you need to either:
hint:
hint: 	git merge --no-ff
hint:
hint: or:
hint:
hint: 	git rebase
hint:
hint: Disable this message with "git config advice.diverging false"
fatal: Not possible to fast-forward, aborting.
```
